src/transformers/models/phi3/modeling_phi3.py

```diff
@@ -1105,21 +1115,30 @@ def _update_causal_mask(
         # to infer the attention mask.
         past_seen_tokens = past_key_values.get_seq_length() if past_key_values is not None else 0
         using_static_cache = isinstance(past_key_values, StaticCache)
+        using_sliding_window_cache = isinstance(past_key_values, SlidingWindowCache)
 
         # When output attentions is True, sdpa implementation's forward method calls the eager implementation's forward
-        if self.config._attn_implementation == "sdpa" and not using_static_cache and not output_attentions:
+        if (
+            self.config._attn_implementation == "sdpa"
+            and not (using_static_cache or using_sliding_window_cache)
+            and not output_attentions
+        ):
             if AttentionMaskConverter._ignore_causal_mask_sdpa(
                 attention_mask,
                 inputs_embeds=input_tensor,
                 past_key_values_length=past_seen_tokens,
+                sliding_window=self.config.sliding_window,
                 is_training=self.training,
             ):
                 return None
 
         dtype, device = input_tensor.dtype, input_tensor.device
         min_dtype = torch.finfo(dtype).min
         sequence_length = input_tensor.shape[1]
-        if using_static_cache:
+        if using_sliding_window_cache:
```

Comment:
we now have a significant code block to determine `target_length`, but `target_length` is not used directly in `_update_causal_mask`

suggestion to avoid this disconnection:
1. make `SlidingWindowCache` store the `config.sliding_window` it receives at init time
2. move the logic that computes `target_length` inside `_prepare_4d_causal_attention_mask_with_cache_position`
3. `_prepare_4d_causal_attention_mask_with_cache_position` no longer receives `target_length` nor `config`, as they can be retrieved from `past_key_values`

Reply:
For the target length, maybe we merge https://github.com/huggingface/transformers/pull/32421/ first where the max length should be easily accessible through cache no matter cache type. We earlier discussed returning max cache shape, and sliding window has a max capacity for cache class

And then yeah, we can move that part to `_prepare_4d_causal_attention_mask_with_cache_position` but I think we'd better move it in all classes for general consistency

Reply:
Agreed with the plan -- shall we leave a TODO for us in this PR, linking to your comment?

Reply:
yes, I'll add a TODO for us so we can make the required change in all models at once, to not mix different updates in one PR :)